CRM/Core/BAO/EntityTag.php

```diff
@@ -483,4 +483,37 @@ public static function buildOptions($fieldName, $context = NULL, $props = []) {
     return $options;
   }
 
+  /**
+   * This function deletes entity tags when a related entity is called.
+   *
+   * It is registered as a listener in \Civi\Core\Container::createEventDispatcher
+   *
+   * Ideally we will later nuance it so it's only called on delete....
+   *
+   * @param \Civi\Core\Event\PreEvent $event
+   */
+  public static function preDeleteOtherEntity($event) {
+    if ($event->action !== 'delete'
+      // Activity can call the pre hook for delete with no ID - this seems to be isolated to activity....
+      // @todo - what is the correct way to standardise?
+      || ($event->entity === 'Activity' && !$event->id)) {
```

Comment:
Wait... what?
Why would this hook be called with no id? Is that a bug?

Reply:
It feels like it to me - but in activity BAO

```
  public static function deleteActivity(&$params, $moveToTrash = FALSE) {
    // CRM-9137
    if (!empty($params['id']) && !is_array($params['id'])) {
      CRM_Utils_Hook::pre('delete', 'Activity', $params['id'], $params);
    }
    else {
      CRM_Utils_Hook::pre('delete', 'Activity', NULL, $params);
    }
```